README.rst
========
pyPostal
========

pyPostal is an Interface for sending real (paper-based) letters via an API.


There are several providers which offer printing, envelope stuffing and posting services but currently only
https://www.pixelletter.de/ provides such services to SME without contractual hassles and the like.

This interface only supports mailing PDFs which have the Address Placed in the PDF at the `DIN 5008 <http://de.wikipedia.org/wiki/DIN_5008>`_ Address Location.

You can download it at the `Python PAckage Index (PyPI) <http://pypi.python.org/pypi/pyPostal/>`_.


High-Level Usage
================

Usage is very easy: Just set up your credentials in the Environment before starting Python::

    export PYPOSTAL_PIXELLETTER_CRED='your@email.com:PASSWORD'
    export PYPOSTAL_SIPGATE_CRED='your@email.com:PASSWORD'

Then call ``pypostal.send_post_pixelletter()`` with the open PDF files or PDF datatream to send and the country code of the recipient::

    >>> import pypostal
    >>> pypostal.send_post_pixelletter(
            [open('/Users/md/Desktop/Testbrief.pdf').read()], 'DE')

For Sending Faxes you need a Sender number 

    >>> import pypostal
    >>> pypostal.send_fax_sipgate(
            [open('/Users/md/Desktop/Testbrief.pdf').read()],
            dest_numbers=['+49123456', '+49654321'], '+49-meine-nummer')


If you prefer to hardcode credentials you can provide them via a function call instead via the environment::

    >>> pypostal.send_post_pixelletter([open('Testbrief.pdf')], 'DE', 
                                       username='your@email.com', 
                                       password='PASSWORD')

In addition the module searches for credentials in the django settings module and in a moduke called `config`.


Pixelletter Interface
=====================

Pixelletter offers a `Bunch of Documentation <https://www.pixelletter.de/de/doku2.php>`_ and a `PHP Library <http://www.pixelletter.de/xml/pixelletter.class.txt>`_. Unfortunately there is no specification of the HTTP-API and the documentation seems also somewhat incomplete and outdated. Also it seems that Pixelletter uses no prebuild XML processing and parsing pipeline but build one arround print statements. This library was build by using trial and error and reverse engeneering the website.


Example Usage
-------------

The Pixelletter interface is straightforward::

    # Log in
    >>> from pypostal import Pixelletter
    >>> pix = Pixelletter('your_email', 'your_password', test_mode=True)
    
    # Show how many Cents Pixelletter owes you.
    >>> print pix.get_account_info()['customer_credit']
    1995
    
    # Send two PDFs from your Desktop as en Letter
    >>> print pix.sendPost([open('/Users/md/Desktop/Testbrief.pdf'), 
                            open('/Users/md/Desktop/Thesis.pdf')])

    # Send one PDF printet in color and in CO2 neutral fashion.
    >>> print pix.sendPost([open('/Users/md/Desktop/Testbrief.pdf').read()], 
                           guid='0815-4711', service=['green', 'color'])

You can provide a GUID ("Transaction Identifier" in the Pixelletter Documenttion) - this might support a Track and Trace Interface but I havn't seen any documentation on this. Something like https://www.pixelletter.de/de/auftraege.php as an `Atom Feed <http://en.wikipedia.org/wiki/Atom_(standard)>`_ vertainly would be nice.

The Python library currently supports following services:

* green (default, use ``service=[]`` to disable)
* einschreiben (see `DHL / Deutsche Post AG <http://www.deutschepost.de/dpag?skin=lo&check=no&lang=de_DE&tab=1&xmlFile=link1015321_6396>`_ on the differences)
* einschreibeneinwurf
* eigenhaendig
* eigenhaendigrueckschein
* rueckschein
* color

The Pixelletter API also seems to support "Nachnahme", "Postident Comfort" and "Ueberweisungsvordruck" but they are undocumented and currently not supported by this library. 


Sipgate.de Fax Interface
========================

Sipgate.de offers a hot and new REST API with `some documentation <http://www.live.sipgate.de/api/rest>`_. Pypostal provides functionality to send PDFs via Fax. You can use `PYPOSTAL_SIPGATE_CRED` to set Credentials.


Example Usage
-------------

A nice and clean interface::

    # Log in
    >>> from pypostal.sipgate import Sipgate
    >>> sip = Sipgate('your_email', 'your_password')
    # Send a PDF via Fax
    >>> sip.sendFax([open('/Users/md/Desktop/Testbrief.pdf')], ['+49123456'], '+49654321')



Planned other Interfaces
========================

We want to support

* Pixelletter Fax interface
* Pawisda L-Vin Post / Pinbriefportal SOAP Interface


Links
=====

* `WWW::Pixelletter <http://cpansearch.perl.org/src/RCL/WWW--Pixelletter-0.1/lib/WWW/Pixelletter.pm>`_ (Perl Module) for Pixelletter
* `PHP Library <http://www.pixelletter.de/xml/pixelletter.class.txt>`_ for Pixelletter
* `Pixelletter Documentation <https://www.pixelletter.de/de/doku2.php>`_
* `pyJasper <http://github.com/hudora/pyJasper>`_ and `iReport <http://www.jaspersoft.com/de/ireport>`_ are a decent way to generate PDFs.


CHANGES
1.1c   uniform reading of credentials, packaging fixed 
1.1    Send Fax via sipgate.de REST API
1.0p2  fixe a distutils issue


pypostal/postal.py
#!/usr/bin/env python
# encoding: utf-8
"""
pypostal/postal.py - sending of letters

Created by Maximillian Dornseif on 2010-08-12.
Copyright (c) 2010 HUDORA. All rights reserved.
"""


import httplib
import mimetypes
import os
import uuid
import xml.etree.ElementTree as ET


def get_content_type(filename):
    return mimetypes.guess_type(filename)[0] or 'application/pdf '


def encode_multipart_formdata(fields, files={}):
    """
    fields is a sequence of (name, value) elements for regular form fields.
    files is a sequence of (name, filename, value) elements for data to be uploaded as files
    Return (content_type, body) ready for httplib.HTTP instance
    """
    # Based on From http://aspn.activestate.com/ASPN/Cookbook/Python/Recipe/146306
    boundary = '----------ThIs_Is_tHe_bouNdaRY_$%s' % (uuid.uuid4())
    out = []
    # if it's dict like then use the items method to get the fields
    for (key, value) in fields.items():
        out.append('--' + boundary)
        out.append('Content-Disposition: form-data; name="%s"' % key)
        out.append('')
        out.append(value)
    for (key, fd) in files.items():
        if hasattr(fd, 'name'):
            filename = fd.name
            filedata = fd.read()
        else:
            filename = key + '.pdf'
            filedata = fd
        out.append('--' + boundary)
        out.append('Content-Disposition: form-data; name="%s"; filename="%s"' % (key, filename))
        out.append('Content-Type: %s' % get_content_type(filename))
        out.append('')
        out.append(filedata)
    out.append('--' + boundary + '--')
    out.append('')
    body = '\r\n'.join([str(x) for x in out])
    content_type = 'multipart/form-data; boundary=%s' % boundary
    return content_type, body


class Pixelletter(object):
    def __init__(self, username, password, test_mode=False):
        self.test_mode = test_mode
        self.username = username
        self.password = password
    
    def POST(self, content_type, content):
        h = httplib.HTTP('www.pixelletter.de')
        h.putrequest('POST', '/xml/index.php')
        h.putheader('host', 'www.pixelletter.de')
        h.putheader('content-type', content_type)
        h.putheader('content-length', str(len(content)))
        h.endheaders()
        h.send(content)
        errcode, errmsg, headers = h.getreply()
        content = h.file.read()
        if str(errcode) != '200':
            raise RuntimeError("%s -- %r" % (errcode, errmsg))
        return content
    
    def _get_auth_xml(self):
        root = ET.Element('pixelletter', version='1.0')
        auth = ET.SubElement(root, 'auth')
        ET.SubElement(auth, 'email').text = self.username
        ET.SubElement(auth, 'password').text = self.password
        ET.SubElement(auth, 'agb').text = 'ja'
        ET.SubElement(auth, 'widerrufsverzicht').text = 'ja'
        ET.SubElement(auth, 'testmodus').text = 'true' if self.test_mode else 'false'
        ET.SubElement(auth, 'ref').text = 'reference#'
        return root
    
    def get_account_info(self):
        """Return a dict with account status information, e.g
        
        {'company': 'Cyberlogi GmbH'
         'customer_credit': 2007, # Betrag in Cent
         'customer_data': '', 
         'customer_id': '353***', 
         'email': '***@cyberlogi.de', 
         'payment_type': 'guthaben', 
         'sex': 'yes', 
         'tel': '184****', 
         'tel_prefix': '017*', 
         'title': 'Dr.', 
         ...
         }
        """
        root = self._get_auth_xml()

        command = ET.SubElement(root, 'command')
        info = ET.SubElement(command, 'info')
        ET.SubElement(info, 'account:info ', type='all')
        data = ET.tostring(root)
        content_type, content = encode_multipart_formdata(fields=dict(xml=data))

        reply = self.POST(content_type, content)
        # Pixelletter's XML is invalid
        # it is messing with namespaces - remove them to generate valid XML
        content = reply.replace('customer:', 'customer_')
        content = content.replace('tel:prefix', 'tel_prefix')
        content = content.replace('fax:prefix', 'fax_prefix')
        content = content.replace('mobil:prefix', 'mobil_prefix')
        content = content.replace('payment:type', 'payment_type')
        response = ET.fromstring(content)
        ret = {}
        for parent in response.getiterator():
            for child in parent:
                text = child.text
                if not text:
                    text = ''
                ret[child.tag] = text.strip()
        ret['customer_credit'] = int(float(ret['customer_credit'])*100)
        return ret


    def sendPost(self, uploadfiles, dest_country='DE', guid='', services=None):
        """Instructs pixelletter.de to send a letter.
            Send one PDF printed in color and in CO2 neutral fashion.
            
            >>> print pix.sendPost(['/Users/md/Desktop/Testbrief.pdf'], guid='0815-4711', service=['green', 'color'])
            
            Uploadfiles is a list of filenames to be send (as a single letter). The first page must contain
            the destination Address.
            
            dest_country is the country where the letter is going to be send to - this is needed for postage calculation.
            
            guid is some tracking ID specific to the user and can be left blank
            
            services is a list of additional services requested. It defaults to ['green']
            The Python library currently supports following services:
            
            * green - GoGreen CO2 neutral postage(default, use ``service=[]`` to disable)
            * einschreiben
            * einschreibeneinwurf
            * eigenhaendig
            * eigenhaendigrueckschein
            * rueckschein
            * color
            """
        if not services:
            services = ['green']
        addoption = set()
        for service in services:
            if service == 'einschreiben':
                addoption.add('27')
            elif service == 'einschreibeneinwurf':
                addoption.add('30')
            elif service == 'eigenhaendig':
                addoption.add('27')
                addoption.add('29')
            elif service == 'eigenhaendigrueckschein':
                addoption.add('27')
                addoption.add('28')
                addoption.add('29')
            elif service == 'rueckschein':
                addoption.add('27')
                addoption.add('29')
            elif service == 'green':
                addoption.add('44')
            elif service == 'color':
                addoption.add('33')
            else:
                raise ValueError('unbekannter servicelevel %s - gueltig ist %r' % (servicelevel,
                    ['einschreiben', 'einschreibeneinwurf', 'eigenhaendig', 'eigenhaendigrueckschein',
                     'rueckschein', 'green', 'color']))
            # Unsupported so far
            # <option value="31" >Nachnahme
            # <option value="42"> Postident Comfort
            # <option value="43" >Überweisungsvordruck
        addoption = ','.join(list(addoption))


        root = self._get_auth_xml()

        order = ET.SubElement(root, 'order', type='upload')
        options = ET.SubElement(order, 'options')
        ET.SubElement(options, 'type').text = 'upload'
        ET.SubElement(options, 'action').text = '1' # Brief
        ET.SubElement(options, 'destination').text = dest_country
        if guid:
            ET.SubElement(options, 'transaction').text = str(guid)
        if addoption:
            ET.SubElement(options, 'addoption').text = addoption

        # Unsupported so far:
        # <fax>
        # <control>
        # <returnaddress>

        data = ET.tostring(root)
        # print data
        form = {'xml': data}
        files = {}
        if len(uploadfiles) < 1:
            raise ValueError( "No files to send...")
        for i, fd in enumerate(uploadfiles):
            files['uploadfile%d' % i] = fd
        content_type, content = encode_multipart_formdata(form, files)
        reply = self.POST(content_type, content)
        return reply if reply else True


def send_post_pixelletter(uploadfiles, dest_country='DE', guid='', services=None, username=None, password=None, test_mode=False):
    if not username:
        os.environ.get('PYPOSTAL_PIXELLETTER_CRED', ':').split(':')[0]
    if not password:
        os.environ.get('PYPOSTAL_PIXELLETTER_CRED', ':').split(':')[1]
    
    pix = Pixelletter(username, password, test_mode=test_mode)
    pix.sendPost(uploadfiles, dest_country, services=services)
    return True


pypostal/sipgate.py
#!/usr/bin/env python
# encoding: utf-8
"""
pypostal/sipgate.py - fax a PDF via the sipgate API

See http://www.live.sipgate.de/api/rest for further description of the sipgate REST API.

Created by Maximillian Dornseif on 2010-08-14.
Copyright (c) 2010 HUDORA. All rights reserved.
"""

import httplib
import os
import urllib
import urlparse

try:
    from django.conf import settings
except ImportError:
    settings = object()
try:
    import config
except:
    config = object()


def add_query(url, params):
    """
    Add GET parameters to a given URL
    
    >>> add_query('/sicrit/', {'passphrase': 'fiftyseveneleven'})
    '/sicrit/?passphrase=fiftyseveneleven'
    """
    
    url_parts = list(urlparse.urlparse(url))
    query = dict(urlparse.parse_qsl(url_parts[4]))
    query.update(params)
    url_parts[4] = urllib.urlencode(query)
    return urlparse.urlunparse(url_parts)


def clean_number(number):
    """Try to clean a phonenumber"""
    
    if number.startswith('0'):
        number = '49' + number[1:]
    number = number.replace(' ', '').replace('-', '').replace('+', '')
    return number


class Sipgate(object):
    def __init__(self, username, password):
        self.version = '2.17.0'
        self.username = username
        self.password = password
        
    @property
    def authheader(self):
        auth = '%s:%s' % (self.username, self.password)
        return "Basic %s" % auth.encode('base64').strip()
    
    def sendFax(self, uploadfiles, dest_numbers, source=None, guid=''):
        if len(uploadfiles) > 1:
            raise ValueError('Sipgate currently only supports single PDF uploading')
        if hasattr(uploadfiles[0], 'name'):
            filedata = uploadfiles[0].read()
        else:
            filedata = uploadfiles[0]
        
        params = {'version': self.version,
                  'targets': ",".join('tel:' + clean_number(dest) for dest in dest_numbers)}
        if source:
            params['source'] = 'tel:' + clean_number(source),
        url = add_query("/my/events/faxes/", params)
        headers = {"Content-Type": "application/pdf", "Authorization": self.authheader}
        
        conn = httplib.HTTPSConnection("api.sipgate.net")
        conn.request("POST", url, headers=headers, body=filedata)
        
        response = conn.getresponse()
        # TODO: Error handling
        # print response.status, response.reason
        # data = response.read()
        # print data
        conn.close()
        return guid


def send_fax_sipgate(uploadfiles, dest_numbers=[], source=None, guid='', username=None, password=None):
    credentials = os.environ.get('PYPOSTAL_SIPGATE_CRED', ':')
    if not credentials:
        credentials = getattr(settings, 'PYPOSTAL_SIPGATE_CRED', None)
    if not credentials:
        credentials = getattr(config, 'PYPOSTAL_SIPGATE_CRED', None)
    if not username:
            username = credentials.split(':')[0]
    if not password:
            password = credentials.split(':')[1]
    
    if (not username) or (not password):
        raise RuntimeError('set PYPOSTAL_SIPGATE_CRED="user:pass"')
    
    sip = Sipgate(username, password)
    return sip.sendFax(uploadfiles, dest_numbers, source)


if __name__ == "__main__":
    import doctest
    doctest.testmod()

pypostal/pixelletter.py
#!/usr/bin/env python
# encoding: utf-8
"""
pypostal/postal.py - sending of letters

Created by Maximillian Dornseif on 2010-08-12.
Copyright (c) 2010 HUDORA. All rights reserved.
"""


import httplib
import mimetypes
import os
import uuid
import xml.etree.ElementTree as ET


try:
    import config
except:
    config = None


def get_content_type(filename):
    return mimetypes.guess_type(filename)[0] or 'application/pdf '


def encode_multipart_formdata(fields, files={}):
    """
    fields is a sequence of (name, value) elements for regular form fields.
    files is a sequence of (name, filename, value) elements for data to be uploaded as files
    Return (content_type, body) ready for httplib.HTTP instance
    """
    # Based on From http://aspn.activestate.com/ASPN/Cookbook/Python/Recipe/146306
    boundary = '----------ThIs_Is_tHe_bouNdaRY_$%s' % (uuid.uuid4())
    out = []
    # if it's dict like then use the items method to get the fields
    for (key, value) in fields.items():
        out.append('--' + boundary)
        out.append('Content-Disposition: form-data; name="%s"' % key)
        out.append('')
        out.append(value)
    for (key, fd) in files.items():
        if hasattr(fd, 'name'):
            filename = fd.name
            filedata = fd.read()
        else:
            filename = key + '.pdf'
            filedata = fd
        out.append('--' + boundary)
        out.append('Content-Disposition: form-data; name="%s"; filename="%s"' % (key, filename))
        out.append('Content-Type: %s' % get_content_type(filename))
        out.append('')
        out.append(filedata)
    out.append('--' + boundary + '--')
    out.append('')
    body = '\r\n'.join([str(x) for x in out])
    content_type = 'multipart/form-data; boundary=%s' % boundary
    return content_type, body


class Pixelletter(object):
    def __init__(self, username, password, test_mode=False):
        self.test_mode = test_mode
        self.username = username
        self.password = password
    
    def POST(self, content_type, content):
        h = httplib.HTTP('www.pixelletter.de')
        h.putrequest('POST', '/xml/index.php')
        h.putheader('host', 'www.pixelletter.de')
        h.putheader('content-type', content_type)
        h.putheader('content-length', str(len(content)))
        h.endheaders()
        h.send(content)
        errcode, errmsg, headers = h.getreply()
        content = h.file.read()
        if str(errcode) != '200':
            raise RuntimeError("%s -- %r" % (errcode, errmsg))
        return content
    
    def _get_auth_xml(self):
        root = ET.Element('pixelletter', version='1.0')
        auth = ET.SubElement(root, 'auth')
        ET.SubElement(auth, 'email').text = self.username
        ET.SubElement(auth, 'password').text = self.password
        ET.SubElement(auth, 'agb').text = 'ja'
        ET.SubElement(auth, 'widerrufsverzicht').text = 'ja'
        ET.SubElement(auth, 'testmodus').text = 'true' if self.test_mode else 'false'
        ET.SubElement(auth, 'ref').text = 'reference#'
        return root
    
    def get_account_info(self):
        """Return a dict with account status information, e.g
        
        {'company': 'Cyberlogi GmbH'
         'customer_credit': 2007, # Betrag in Cent
         'customer_data': '', 
         'customer_id': '353***', 
         'email': '***@cyberlogi.de', 
         'payment_type': 'guthaben', 
         'sex': 'yes', 
         'tel': '184****', 
         'tel_prefix': '017*', 
         'title': 'Dr.', 
         ...
         }
        """
        root = self._get_auth_xml()

        command = ET.SubElement(root, 'command')
        info = ET.SubElement(command, 'info')
        ET.SubElement(info, 'account:info ', type='all')
        data = ET.tostring(root)
        content_type, content = encode_multipart_formdata(fields=dict(xml=data))

        reply = self.POST(content_type, content)
        # Pixelletter's XML is invalid
        # it is messing with namespaces - remove them to generate valid XML
        content = reply.replace('customer:', 'customer_')
        content = content.replace('tel:prefix', 'tel_prefix')
        content = content.replace('fax:prefix', 'fax_prefix')
        content = content.replace('mobil:prefix', 'mobil_prefix')
        content = content.replace('payment:type', 'payment_type')
        response = ET.fromstring(content)
        ret = {}
        for parent in response.getiterator():
            for child in parent:
                text = child.text
                if not text:
                    text = ''
                ret[child.tag] = text.strip()
        ret['customer_credit'] = int(float(ret['customer_credit'])*100)
        return ret


    def sendPost(self, uploadfiles, dest_country='DE', guid='', services=None):
        """Instructs pixelletter.de to send a letter.
            Send one PDF printed in color and in CO2 neutral fashion.
            
            >>> print pix.sendPost(['/Users/md/Desktop/Testbrief.pdf'], guid='0815-4711', service=['green', 'color'])
            
            Uploadfiles is a list of filenames to be send (as a single letter). The first page must contain
            the destination Address.
            
            dest_country is the country where the letter is going to be send to - this is needed for postage calculation.
            
            guid is some tracking ID specific to the user and can be left blank
            
            services is a list of additional services requested. It defaults to ['green']
            The Python library currently supports following services:
            
            * green - GoGreen CO2 neutral postage(default, use ``service=[]`` to disable)
            * einschreiben
            * einschreibeneinwurf
            * eigenhaendig
            * eigenhaendigrueckschein
            * rueckschein
            * color
            """
        if not services:
            services = ['green']
        addoption = set()
        for service in services:
            if service == 'einschreiben':
                addoption.add('27')
            elif service == 'einschreibeneinwurf':
                addoption.add('30')
            elif service == 'eigenhaendig':
                addoption.add('27')
                addoption.add('29')
            elif service == 'eigenhaendigrueckschein':
                addoption.add('27')
                addoption.add('28')
                addoption.add('29')
            elif service == 'rueckschein':
                addoption.add('27')
                addoption.add('29')
            elif service == 'green':
                addoption.add('44')
            elif service == 'color':
                addoption.add('33')
            else:
                raise ValueError('unbekannter servicelevel %s - gueltig ist %r' % (servicelevel,
                    ['einschreiben', 'einschreibeneinwurf', 'eigenhaendig', 'eigenhaendigrueckschein',
                     'rueckschein', 'green', 'color']))
            # Unsupported so far
            # <option value="31" >Nachnahme
            # <option value="42"> Postident Comfort
            # <option value="43" >Überweisungsvordruck
        addoption = ','.join(list(addoption))


        root = self._get_auth_xml()

        order = ET.SubElement(root, 'order', type='upload')
        options = ET.SubElement(order, 'options')
        ET.SubElement(options, 'type').text = 'upload'
        ET.SubElement(options, 'action').text = '1' # Brief
        ET.SubElement(options, 'destination').text = dest_country
        if guid:
            ET.SubElement(options, 'transaction').text = str(guid)
        if addoption:
            ET.SubElement(options, 'addoption').text = addoption

        # Unsupported so far:
        # <fax>
        # <control>
        # <returnaddress>

        data = ET.tostring(root)
        # print data
        form = {'xml': data}
        files = {}
        if len(uploadfiles) < 1:
            raise ValueError( "No files to send...")
        for i, fd in enumerate(uploadfiles):
            files['uploadfile%d' % i] = fd
        content_type, content = encode_multipart_formdata(form, files)
        reply = self.POST(content_type, content)
        if not '<msg>Auftrag erfolgreich ' in reply:
            raise RuntimeError("API fehler: %s" % (reply))
        return guid, ''


def send_post_pixelletter(uploadfiles, dest_country='DE', guid='', services=None, username=None,
                          password=None, test_mode=False):
    credentials = os.environ.get('PYPOSTAL_PIXELLETTER_CRED', ':')
    if not credentials:
        credentials = getattr(settings, 'PYPOSTAL_PIXELLETTER_CRED', None)
    if not credentials:
        credentials = getattr(config, 'PYPOSTAL_PIXELLETTER_CRED', None)
    if not username:
            username = credentials.split(':')[0]
    if not password:
            password = credentials.split(':')[1]

    if (not username) or (not password):
        raise RuntimeError('set PYPOSTAL_PIXELLETTER_CRED="user:pass"')
    pix = Pixelletter(username, password, test_mode=test_mode)
    return pix.sendPost(uploadfiles, dest_country, services=services)
